vortex-array/src/array/chunked/compute/filter.rs

```diff
@@ -0,0 +1,240 @@
+    }
+
+    // Pad the end of the buffer with false, if necessary.
+    let n_trailing_false = len - pos;
+    buffer.append_n(n_trailing_false, false);
+
+    BoolArray::from(buffer.finish()).into_array()
+}
+
+/// Filter the chunks using slice ranges.
+fn filter_slices<'a>(
+    array: &'a ChunkedArray,
+    set_slices: Box<dyn Iterator<Item = (usize, usize)> + 'a>,
+) -> VortexResult<Vec<Array>> {
+    let mut result = Vec::with_capacity(array.nchunks());
+
+    // Pre-materialize the chunk ends for performance.
+    // The chunk ends is nchunks+1, which is expected to be in the hundreds or at most thousands.
+    let chunk_ends = array.chunk_offsets().into_canonical()?.into_primitive()?;
+    let chunk_ends = chunk_ends.maybe_null_slice::<u64>();
+
+    let mut chunk_filters = vec![ChunkFilter::None; array.nchunks()];
+
+    for (slice_start, slice_end) in set_slices {
+        let (start_chunk, start_idx) = find_chunk_idx(slice_start, chunk_ends);
+        // NOTE: we adjust slice end back by one, in case it ends on a chunk boundary, we do not
+        // want to index into the unused chunk.
+        let (end_chunk, end_idx) = find_chunk_idx(slice_end - 1, chunk_ends);
+        // Adjust back to an exclusive range
+        let end_idx = end_idx + 1;
+
+        if start_chunk == end_chunk {
+            // start == end means that the slice lies within a single chunk.
+            match &mut chunk_filters[start_chunk] {
+                f @ (ChunkFilter::All | ChunkFilter::None) => {
+                    *f = ChunkFilter::Slices(vec![(start_idx, end_idx)]);
+                }
+                ChunkFilter::Slices(slices) => {
+                    slices.push((start_idx, end_idx));
+                }
+            }
+        } else {
+            // start != end means that the range is split over at least two chunks:
+            // start chunk: append a slice from (start_idx, start_chunk_end).
+            // end chunk: append a slice from (0, end_idx).
+            // chunks between start and end: append ChunkFilter::All.
+            let start_chunk_end = chunk_ends[start_chunk + 1];
+            let start_slice = (start_idx, start_chunk_end as _);
+            match &mut chunk_filters[start_chunk] {
+                f @ (ChunkFilter::All | ChunkFilter::None) => {
+                    *f = ChunkFilter::Slices(vec![start_slice])
+                }
+                ChunkFilter::Slices(slices) => slices.push(start_slice),
+            }
+
+            let end_slice = (0, end_idx);
+            match &mut chunk_filters[end_chunk] {
+                f @ (ChunkFilter::All | ChunkFilter::None) => {
+                    *f = ChunkFilter::Slices(vec![end_slice]);
+                }
+                ChunkFilter::Slices(slices) => slices.push(end_slice),
+            }
+
+            #[allow(clippy::needless_range_loop)]
+            for chunk in (start_chunk + 1)..end_chunk {
+                chunk_filters[chunk] = ChunkFilter::All;
+            }
```

Comment:
one thing that the clippy suggestion does is avoid bounds checking but it would only show up on large arrays

Reply:
I forgot that you can do `&mut vector[start...end]`, which I think is both clear and avoids bounds check